Make `asString` read any scalar config type

ruippeixotog opened the pull request:
Makes `asString` read any scalar config type, just like the Typesafe Config API does. Since the underlying config format allows strings to be defined without quotes and primitives like `true` and `null` can arguably be seen as strings, this "cast" makes sense to me in PureConfig.

Fixes #338.

ruippeixotog committed:
Make `asString` read any scalar config type

leifwickland reviewed (approved):
Seems plausible.

jcazevedo reviewed (changes requested):
I think we want `ConfigCursor` behaviour regarding casts to be similar to what Typesafe Config does. As such, I'm not sure we should be supporting `null` as `String`, since Typesafe Config [doesn't](https://github.com/lightbend/config/blob/master/config/src/main/java/com/typesafe/config/impl/DefaultTransformer.java#L60).

ruippeixotog commented:
@jcazevedo that's right, I read the code wrong, I added support for `null` thinking Typesafe Config would convert it.

The comment explaining their decision seems to be not much about semantics, but about the inability for them to distinguish a null from a missing value on their API:

```
// if we converted null to string here, then you wouldn't properly
// get a missing-value error if you tried to get a null value
// as a string.
```

However, I think I prefer to keep the behavior close to Typesafe Config. Readers can always test if the cursor `isNull` before casting `asString` and handle the special case manually.

ruippeixotog committed:
Do not cast null values to strings in `asString`

ruippeixotog committed:
Merge branch 'master' into improve-cursor-asstring

ruippeixotog committed:
Do not cache .rvm folder on Travis
